app/services/sandbox_url_updator.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+class SandboxUrlUpdator
+  class InvalidUrlError < StandardError; end
+  class MismatchedWikiError < StandardError; end
+  class UserNotExist < StandardError; end
+
+  def initialize(new_url = nil, assignment = nil)
+    @new_url = new_url
+    @assignment = assignment
+  end
+
+  def update
+    validate_new_url
+    @assignment.update_sandbox_url(@new_url)
+    return { partial: 'updated_assignment', locals: { assignment: @assignment } }
+  rescue StandardError => e
+    case e
+    when SandboxUrlUpdator::InvalidUrlError
+      { json: { errors: e, message: e.message }, status: :bad_request }
+    when SandboxUrlUpdator::MismatchedWikiError,
+         SandboxUrlUpdator::UserNotExist
+      { json: { errors: e, message: e.message }, status: :unprocessable_entity }
+    else
+      { json: { errors: e, message: e.message }, status: :internal_server_error }
+    end
+  end
+
+  def validate_new_url
+    existing_url = @assignment.sandbox_url
+    existing_language, existing_project = existing_url.match(%r{https://([^.]*)\.([^.]*)\.org}).captures
+    new_url_match = @new_url.match(%r{^https://([^./]++)\.([^./]++)\.org/wiki/User:([^/]+)/([^/#<>\[\]|{}:.]+)(?<!\.{3})(?<!\.)/?$})
+    # Handle invalid url
+    raise InvalidUrlError, I18n.t('assignments.invalid_url', url: @new_url) unless new_url_match
+    # Handle mismatched wiki
+    new_language, new_project, new_username = new_url_match.captures
+    unless existing_language == new_language && existing_project == new_project
+      raise MismatchedWikiError, I18n.t('assignments.mismatched_wiki', url: @new_url)
+    end
+    # Handle user not exist
+    new_user = User.find_by(username: new_username)
+    raise UserNotExist, I18n.t('assignments.no_user_exists', username: new_username) unless new_user
```

Comment:
I'm not sure why this raised an error in my testing even for a user that does exist, but I don't think a check for a user is necessary or helpful anyway. I would say checking that it is a non-empty title that includes the `User:` prefix is good enough.

Reply:
sure, removed that check and done the same